### PR TITLE
First version of test workflow document to joint-development repository

### PR DESCRIPTION
**Pull request by pekkaan**

### CODE REVIEW

File: TEST_WORKFLOW.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+1.4 Adherence to this policy are made legally binding to Vendors by inclusion of appropriate stipulations in contracts between Partners and Vendors. Vendor in context of this policy is a firm or other organisation performing development work by assigment of a Partner.
+
+1.5 Partners and Developers implement this policy in good faith and in the context of sustainable, good software development practice.
+
+1.6 Projects may have their own workflow arrangements as far as these do not contradict this policy.
+
+## 2 Related documents
+
+2.1 Workflow policy is related to other X-Road tests development policy and regulations:
+- workflow policy implements Harmonized X-road tests document [Harmonized X-Road test environment](https://github.com/ria-ee/blob/master/HARMONIZED_TEST_ENVIRONMENT.md)
```
Comment (hanhaka) [inline TEST_WORKFLOW.md]: Link does not work, "Not found" error?

File: TEST_WORKFLOW.md [hunk 2]
```diff
@@ -0,0 +1,136 @@
+X-Road Tests development
+# Workflow Policy
+
+v1.0
+
+## 1	General
+
+1.1	This document establishes workflow policy for joint development of X-Road software. Workflow Policy strives to serve as a handbook for all practical aspects of development.
```
Comment (hanhaka) [inline TEST_WORKFLOW.md]: This document establishes workflow policy for joint development of X-Road software _test development_.

File: TEST_WORKFLOW.md (hunk 1 above)
Comment (iluwatar) [inline TEST_WORKFLOW.md]: I think it's incorrect to say that workflow implements harmonized X-Road tests document. Rather the harmonized document could just be listed here. Also the link doesn't work.

File: TEST_WORKFLOW.md [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+4.1	The following code and documentation repositories are used:
+- Master Test Repository - short name: X-Road-tests; hosted by: GitHub; managed by: RIA; purpose: release of X-Road automated tests; access: Head Architect has write access; read access: ALL.
+	- https://github.com/ria-ee/X-Road-tests
+
+4.2	Partners can establish their own, additional repositories, for backup, software distribution or other purposes.
+
+## 5	Branching pattern
+
+5.1	Branching pattern follows the [Gitflow model](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow). Two perpetual branches – `master` , `develop` and own `testfeature` branches.
```
Comment (iluwatar) [inline TEST_WORKFLOW.md]: There is still `testfeature` branch. It should be just `feature` branch.

File: TEST_WORKFLOW.md [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+- `master` branch is used to release X-Road tests
+- `develop` branch is used to test development
```
Comment (iluwatar) [inline TEST_WORKFLOW.md]: Grammatical fix, is used to -> is used for

File: TEST_WORKFLOW.md [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+Main repository:
+- `master` branch https://github.com/ria-ee/X-Road-tests/tree/master
+	- releases of X-Road-tests that correspond to a specific release of X-Road source code. Several versions of X-Road-tests may point to the same X-Road source code version if necessary.
+	-  `master` branch releases are tagged with e.g "X-Road-tests for X-Road 6.16.0"
+- `develop` branch https://github.com/ria-ee/X-Road-tests/tree/develop
+	- Integration branch between Finland and Estonia (and other possible partners)
+	- Latest integrated test development source code 
+
+README.md file in the root of X-Road-tests must indicate the X-Road source code version for both MASTER and DEVELOP branches so that there is no confusion for which version these tests were designated for. Indication is done by pointing to the specific tag (MASTER) or commit (DEVELOP) of the X-Road source code.
```
Comment (iluwatar) [inline TEST_WORKFLOW.md]: Why this new convention (MASTER) and (DEVELOP). Elsewhere it is just `master` and `develop` and it should be used here also for consistency.

File: TEST_WORKFLOW.md [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+1.2	Workflow is understood as systematic arrangement of work, a complex of processes, practices, roles and responsibilities, communication patterns, and artefacts.
+
+1.3	Goals of the workflow policy are to
+- establish productive and secure collaborative open source working environment
+- assure production of high quality software
+- avoid duplication of effort, facilitate re-use of software
+- transparency and openness
+- broader community of developers
+- use of software development best practice
+- clear communication among Partners as well as other stakeholders in the X-Road development process
```
Comment (hanhaka) [inline TEST_WORKFLOW.md]: ...stakeholders in the X-Road _test_ development process

File: TEST_WORKFLOW.md [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+## 7 Changelog management and Pull requests
+
+Changelog description is visible in pull requests. Pull requests from Partners made against the `ria-ee/X-Road-tests/master` or `ria-ee/X-Road-tests/develop` branch MUST follow these conventions:
+
+Pull request name format is 'country-month-year-description':
+Country=origin of the pull request
+Month=Month when the pull request was created
+Year=Year when the pull request was created
+Description=(short) Description of the pull request
+
```
Comment (hanhaka) [inline TEST_WORKFLOW.md]: In HTML, all the text/words are written together:
Pull request name format is 'country-month-year-description': Country=origin of the pull request Month=Month when the pull request was created Year=Year when the pull request was created Description=(short) Description of the pull request

Should be a bulleted list for better reading.

File: TEST_WORKFLOW.md [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+8.2.5 Pull request
+- Pull requests are generally reviewed and accepted on first-come, first-served (FCFS) basis.Non-functional requirements
+- Descriptive git commit messages
+	- E.g "Added xroad-global-configuration test case 3.3" or "Fixed login test working with new layout"
+
+
+8.2.6 Browser acceptance testing
+- Browser versions
+    - Mozilla firefox versionin tests README.md
```
Comment (hanhaka) [inline TEST_WORKFLOW.md]: versionin => version

File: TEST_WORKFLOW.md [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+	
+	
+
+Fork repository:
+- `master` branch https://github.com/vrk-kpa/X-Road-tests and https://github.com/asaquality/X-Road-tests
+	- Tagged releases of X-road-tests
+- `develop` branch https://github.com/vrk-kpa/X-Road-tests and https://github.com/asaquality/X-Road-tests 
+	- Country branch for test development
+	- Latest country test development source code 
+	- Make pull request to develop for ria-ee/X-Road-tests
```
Comment (iluwatar) [inline TEST_WORKFLOW.md]: I think the purpose of the fork's branches is already clear (the same as in the master repository). Instead we could mention the [forking workflow](https://www.atlassian.com/git/tutorials/comparing-workflows#forking-workflow) as the way the repositories are arranged.

File: TEST_WORKFLOW.md [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+
+8.2.6 Browser acceptance testing
+- Browser versions
+    - Mozilla firefox versionin tests README.md
+    - Chrome version version tests README.md
+    	- Chrome used only smoke tests (excluding file download and upload, and certificates)
+
+## 9 Bug fixes
+
+When an error is detected, the Partner who located the error is responsible of detecting the source of the error. Detecting the source of the error means tracking the first version of the software that contains the error. When the source is detected the Partner who created the version in question must be notified. The Partner who created the version is responsible of fixing the problem.
```
Comment (triinum) [inline TEST_WORKFLOW.md]: Comlicated sentence, better make it: When a partner located an error, they are responsible of detecting its source and tracking the first version of the software that contains the error. In order to fix the problem, the Partner who created the version in question must be notified.

File: TEST_WORKFLOW.md [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+8.2.1 Source code
+
+- Is the source code for tests and its dependencies available?
+
+8.2.2 CI build & tests
+- No merge conflicts
+- Test must be working LXC jenkins environment
+- Jenkins job is generated or added existing one
+- Test must working reliably e.g executed multiple times
+- If possible to make test independent
```
Comment (triinum) [inline TEST_WORKFLOW.md]: Grammatical fix: If possible test must work independently

File: TEST_WORKFLOW.md [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+
+4.2	Partners can establish their own, additional repositories, for backup, software distribution or other purposes.
+
+## 5	Branching pattern
+
+5.1	Branching pattern follows the [Gitflow model](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow). Two perpetual branches – `master` , `develop` and own `testfeature` branches.
+
+- `master` branch is used to release X-Road tests
+- `develop` branch is used to test development
+- `feature` branches are used to work on test features
```
Comment (hanhaka) [inline TEST_WORKFLOW.md]: `feature` branches => `testfeature` branches

File: TEST_WORKFLOW.md [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+8.2	The pull request is reviewed by Finland and Estonia responsible persons. Additional reviewers can be added as necessary. The pull request is reviewed according to the acceptance criteria that was in effect when the work on this changeset started.
+
+8.2.1 Source code
+
+- Is the source code for tests and its dependencies available?
+
+8.2.2 CI build & tests
+- No merge conflicts
+- Test must be working LXC jenkins environment
+- Jenkins job is generated or added existing one
```
Comment (iluwatar) [inline TEST_WORKFLOW.md]: What does it mean "Jenkins job is generated or added existing one"?

File: TEST_WORKFLOW.md [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+
+8.2	The pull request is reviewed by Finland and Estonia responsible persons. Additional reviewers can be added as necessary. The pull request is reviewed according to the acceptance criteria that was in effect when the work on this changeset started.
+
+8.2.1 Source code
+
+- Is the source code for tests and its dependencies available?
+
+8.2.2 CI build & tests
+- No merge conflicts
+- Test must be working LXC jenkins environment
```
Comment (iluwatar) [inline TEST_WORKFLOW.md]: Grammatical fix, Test must be working _in_ LXC jenkins environment

File: TEST_WORKFLOW.md (hunk 5 above)
Comment (hanhaka) [inline TEST_WORKFLOW.md]: Change formatting of following words for better reading and analogue: MASTER => `master`, DEVELOP => `develop`, (MASTER) => (`master`) and (DEVELOP) => (`develop`)

File: TEST_WORKFLOW.md [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+
+Changelog description is visible in pull requests. Pull requests from Partners made against the `ria-ee/X-Road-tests/master` or `ria-ee/X-Road-tests/develop` branch MUST follow these conventions:
+
+Pull request name format is 'country-month-year-description':
+Country=origin of the pull request
+Month=Month when the pull request was created
+Year=Year when the pull request was created
+Description=(short) Description of the pull request
+
+e.g Finnish-08-2017-Description-of-pull-request, Estonian-07-2017-Description-of-pull-request
```
Comment (hanhaka) [inline TEST_WORKFLOW.md]: For better reading and to show that this is example text, use Italic text:
 e.g _Finnish-08-2017-Description-of-pull-request_, _Estonian-07-2017-Description-of-pull-request_

File: TEST_WORKFLOW.md [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+- Changelog:
+    - Version information
+    - Notes for changes
+
+8.2.4 Licensing
+- License exists in root folder
+- Is the licensing of the tests and its dependencies ok?
+
+8.2.5 Pull request
+- Pull requests are generally reviewed and accepted on first-come, first-served (FCFS) basis.Non-functional requirements
```
Comment (hanhaka) [inline TEST_WORKFLOW.md]: Non-functional requirements? Something is missing here...?

File: TEST_WORKFLOW.md [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+- Descriptive git commit messages
+	- E.g "Added xroad-global-configuration test case 3.3" or "Fixed login test working with new layout"
+
+
+8.2.6 Browser acceptance testing
+- Browser versions
+    - Mozilla firefox versionin tests README.md
+    - Chrome version version tests README.md
+    	- Chrome used only smoke tests (excluding file download and upload, and certificates)
```
Comment (iluwatar) [inline TEST_WORKFLOW.md]: 8.2.5 and 8.2.6 aren't pull request acceptance criterion anymore, right? I think they should be 8.3 and 8.4.

File: TEST_WORKFLOW.md [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+    - Notes for changes
+
+8.2.4 Licensing
+- License exists in root folder
+- Is the licensing of the tests and its dependencies ok?
+
+8.2.5 Pull request
+- Pull requests are generally reviewed and accepted on first-come, first-served (FCFS) basis.Non-functional requirements
+- Descriptive git commit messages
+	- E.g "Added xroad-global-configuration test case 3.3" or "Fixed login test working with new layout"
```
Comment (hanhaka) [inline TEST_WORKFLOW.md]: For better reading and to show that this is example text, use Italic text here:
_"Added xroad-global-configuration test case 3.3"_ or _"Fixed login test working with new layout"_

**Review by hanhaka — CHANGES_REQUESTED** [review]

File: TEST_WORKFLOW.md [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+
+
+8.2.6 Browser acceptance testing
+- Browser versions
+    - Mozilla firefox versionin tests README.md
+    - Chrome version version tests README.md
+    	- Chrome used only smoke tests (excluding file download and upload, and certificates)
+
```
Comment (iluwatar) [inline TEST_WORKFLOW.md]: What happens when defects are found in the pull request review? And the other case what happens when the pull request is accepted?

File: TEST_WORKFLOW.md (hunk 11 above)
Comment (pekkaan) [reply TEST_WORKFLOW.md]: done

File: TEST_WORKFLOW.md (hunk 2 above)
Comment (pekkaan) [reply TEST_WORKFLOW.md]: done

File: TEST_WORKFLOW.md (hunk 6 above)
Comment (pekkaan) [reply TEST_WORKFLOW.md]: done

File: TEST_WORKFLOW.md (hunk 1 above)
Comment (pekkaan) [reply TEST_WORKFLOW.md]: will be working in future, that should be final location. After test PR is merged.

File: TEST_WORKFLOW.md (hunk 12 above)
Comment (pekkaan) [reply TEST_WORKFLOW.md]: i guess Ilkka comment last round, that could just "feature" branch

File: TEST_WORKFLOW.md (hunk 5 above)
Comment (pekkaan) [reply TEST_WORKFLOW.md]: done

File: TEST_WORKFLOW.md (hunk 7 above)
Comment (pekkaan) [reply TEST_WORKFLOW.md]: changed

File: TEST_WORKFLOW.md (hunk 15 above)
Comment (pekkaan) [reply TEST_WORKFLOW.md]: done

File: TEST_WORKFLOW.md (hunk 18 above)
Comment (pekkaan) [reply TEST_WORKFLOW.md]: done

File: TEST_WORKFLOW.md (hunk 8 above)
Comment (pekkaan) [reply TEST_WORKFLOW.md]: done

File: TEST_WORKFLOW.md (hunk 1 above)
Comment (pekkaan) [reply TEST_WORKFLOW.md]: Added just document. Link will be working when other PR' is accepted

File: TEST_WORKFLOW.md (hunk 3 above)
Comment (pekkaan) [reply TEST_WORKFLOW.md]: changed

File: TEST_WORKFLOW.md (hunk 4 above)
Comment (pekkaan) [reply TEST_WORKFLOW.md]: done

File: TEST_WORKFLOW.md (hunk 5 above)
Comment (pekkaan) [reply TEST_WORKFLOW.md]: changed according to Hannu's comment

File: TEST_WORKFLOW.md (hunk 14 above)
Comment (pekkaan) [reply TEST_WORKFLOW.md]: done

File: TEST_WORKFLOW.md (hunk 11 above)
Comment (pekkaan) [reply TEST_WORKFLOW.md]: done

File: TEST_WORKFLOW.md (hunk 10 above)
Comment (pekkaan) [reply TEST_WORKFLOW.md]: Yes it is complicated. Copy pasted from X-Road develop. It is better to keep it same X-Road develop project. Based on earlier Hannu's ja Ilkka's comment

File: TEST_WORKFLOW.md (hunk 16 above)
Comment (pekkaan) [reply TEST_WORKFLOW.md]: Non-functional requirements added in different PR'. Do it next round

File: TEST_WORKFLOW.md (hunk 9 above)
Comment (pekkaan) [reply TEST_WORKFLOW.md]: Changed bit, but still discussed that will be leaved to document as more clarification.

File: TEST_WORKFLOW.md (hunk 17 above)
Comment (pekkaan) [reply TEST_WORKFLOW.md]: Yes re-ordered Browser to 8.2.5 and moved 8.2.5 -> 8.3

File: TEST_WORKFLOW.md (hunk 19 above)
Comment (pekkaan) [reply TEST_WORKFLOW.md]: Yes, that is missing, Added three chapters for this.

File: TEST_WORKFLOW.md (hunk 13 above)
Comment (pekkaan) [reply TEST_WORKFLOW.md]: Changed to "Tests must be attached to existing or new jenkins task"

**Review by VitaliStupin — APPROVED** [review]

**Review by iluwatar — APPROVED** [review]

**Review by hanhaka — APPROVED** [review]

**Pull request merged**